### PR TITLE
38 inference on single file

### LINKED ISSUE #38: inference on single file
Ok Ellie I changed my mind: I think we should treat the inference logic inside the AudioClassifier by adding some additional methods to the [BaseAudioClassifier](https://github.com/magcil/deepaudio-x/blob/07fadbc52a8cb7d1a602240f6a1c9ca940709023/src/deepaudiox/modules/base_audio_classifier.py#L9) class instead of creating separate scripts inside `src/deepaudiox/inference`.

To this end, we want to add the following two methods on the `BaseAudioClassifier` class:

- `inference_on_waveform`
- `inference_on_path`

Essentially, `inference_on_waveform` will do the job and `inference_on_path` should call `inference_on_waveform`. Below I explain these two methods in detail:

**`inference_on_waveform`**: This method will get a waveform (`y`) as input, the `sample_rate` of the input waveform, the `class_mapping` corresponding to the dictionary mapping the integers to the class_names, and an optional parameter `segment_duration` that will process the audio into chunks of length `segment_duration`. The default value of `segment_duration` should be set as `None`, where in this case the whole audio waveform will be processed as is (without chunking). The function should return an `AudioPrediction` object defined as:

```python
from dataclasses import dataclass
import numpy as np

@dataclass
class AudioPrediction:
    final_label: str
    final_posterior: float
    final_logits: np.ndarray

    segment_labels: list[str] | None = None
    segment_posteriors: list[float] | None = None
    segment_logits: np.ndarray | None = None
```

Store this AudioPrediction object in `srd/deepaudiox/dtos/dataset_items.py`.

**Implementation details**: To implement `inference_on_waveform` use the predefined method [`predict`](https://github.com/magcil/deepaudio-x/blob/07fadbc52a8cb7d1a602240f6a1c9ca940709023/src/deepaudiox/modules/base_audio_classifier.py#L37). In case no `segment_duration` is given, then simply use the predict method on the whole input waveform to get the `y_preds`, corresponding the class_index of the predicted class. Use the `class_mapping` to convert this index to a class_name. Then, the `final_label` should contain this label, `final_posterior` should contain the posterior, and `final_logits` should contain the corresponding logit value. All other parameters (i.e., `segment_*`) should be `None` in this case. Now, if a segment_duration is given, then you should chunk the waveform into the predetermined segments and create a batch consisting of these segments. Use again `predict` method to get the outputs. Then, use majority vote to get the `final_label`, `final_posterior` should the mean value of the posteriors corresponding to the winner, and `final_logits` should be the mean value of the logits corresponding to the winner. In this case, `segment_labels` should contain the list of the labels for each segment, in the same manner `segment_posteriors, and `segment_logits` should contain the list of posteriors, and logits for each segment.

**inference_on_file**: This will get as input the `filename` corresponding to the absolute path of the audio file and will use `inference_on_waveform` to get the results and return an `AudioPrediction` object as well. Use librosa to read the audio file.

Thats all! :) 

Feel free to post here or ask me on DM if you have any questions 🚀

### PR DESCRIPTION
**Pull request by ellievak**
@stefanos-vlachos @ChrisNick92 

This PR adds two methods to the `BaseAudioClassifier` class in [baseclasses.py](https://github.com/magcil/deepaudio-x/blob/38-inference-on-single-file/src/deepaudiox/modules/baseclasses.py): 
- inference_on_waveform 
- inference_on_file 

These methods enable running model inference directly on a waveform or by loading an audio file.

In both cases, inference can be performed either on the entire waveform/file or on fixed-duration segments specified by the user. When segment-level inference is used, both a final aggregated prediction and each segment's predictions, along with relevant posteriors, are returned.

When segments are used, a prediction is first obtained for each segment. Then, the final label is determined based on majority vote on segment predictions. In case of a tie, the final label will be the tied class with the highest mean posterior across all its corresponding segments. The final posterior returned corresponds to the mean posterior of the final label across all segments predicted as that label.

To test the functionality of these methods:

```py
# Inference on a waveform

prediction = model.inference_on_waveform(
    x=waveform,
    sample_rate=sample_rate,
    class_mapping=class_mapping,
    segment_duration=segment_duration  # segment duration in sec, or None for inference on entire waveform
)

print(prediction.final_label)
print(prediction.final_posterior)

print(prediction.segment_labels)  # if segments are used for inference, else these are set to None
print(prediction.segment_posteriors)


# Inference on an audio file (WAV/MP3)

audio_path = "path/to/example.wav"

prediction = model.inference_on_file(
    path=audio_path,
    sample_rate=sample_rate,
    class_mapping=class_mapping,
    segment_duration=segment_duration  # or None, just like in inference_on_waveform
)

print(prediction.final_label)
print(prediction.final_posterior)

print(prediction.segment_labels)
print(prediction.segment_posteriors)
```

### CODE REVIEW

**Review by ChrisNick92 — APPROVED** [review]
Ok Ellie the logic was correct, I slightly changed the logic on `inference_on_waveform` method to reduce the if-else statements and refine the overall workflow of the method. Now, the last segment is not discarded if the total_duration is not divisible by the segment duration. The current behavior pads the remaining samples with 0 and all segments form a batch that are processed by the model.

I merge this PR as everything runs smoothly.
